### ISSUE TITLE
Exposed ports become unresponsive after heavy load

### ISSUE THREAD

**Issue by micdah** [docker/for-win#426]
Michael Friis directed me to submit an issue here (_see [issue 30400](https://github.com/docker/docker/issues/30400) for more_)

I am experiencing an intermittent issue with Docker for Windows, where suddenly all the exposed ports become unresponsive, no connection can be made to the containers. This happens when a lot of activity is put on the containers from the host machine, I am running 4 containers and 11 services on the host machine as well as a handful of websites and API's which all interact with the containers.

### How to reproduce
As requested by Michael Friis, I have made some sample code which seems to be able to reproduce the issue. You can see and clone the code here [github.com/micdah/DockerBomb](https://github.com/micdah/DockerBomb/). I have also made a YouTube video where I demonstrate the issue using my sample code [youtube.com/watch?v=v5k1D60h0zE](https://www.youtube.com/watch?v=v5k1D60h0zE)

I have described how to use the program in the _readme.md_ file in the github repo. **Note that it might take anywhere from a few minutes to minutes before the issue triggers, it is somewhat random - likely because it is tightly timing related**

The sample program creates the requested number of threads, each creating a single connection to the redis container and issuing as many commands as possible until the connection fails.

As demonstrated, when the issue has occurred the container becomes unresponsive on the exposed ports, although it is still running. Trying to restart the container results in an input/output error when trying to bind to the host port. In my previous issue report (_30400_) I have also included a ```netstat``` dump to show that it is not because the port is reserved, when trying to restart the container, that it fails.

### Expected behavior
I would expect the container to continue to be accessible via the exposed ports, as long as it is running. If some resource pool (_handles, connection pool, etc._) is exhausted, I would expect the container to become responsive again when the resources become available again (_for example when stopping the heavy load on the container_).

### Information
**Diagnostic ID**
This is a diagnostic uploaded, just after the issue has occurred, reproduced as described above.
```
30667474-C49F-4185-B957-3A7AE1F38393/2017-01-24_21-44-30
```

**Output of ```docker version```**
```
Client:
 Version:      1.13.0
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Wed Jan 18 16:20:26 2017
 OS/Arch:      windows/amd64

Server:
 Version:      1.13.0
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Wed Jan 18 16:20:26 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of ```docker info```**
```
Containers: 5
 Running: 1
 Paused: 0
 Stopped: 4
Images: 6
Server Version: 1.13.0
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 2f7393a47307a16f8cee44a37b262e8b81021e3e
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.4-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.837 GiB
Name: moby
ID: 5DLJ:7BM4:KTMA:L5UV:ACM5:HJQP:V2W3:ZQXJ:LUS5:XEVE:FJK2:KH5K
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 21
 Goroutines: 28
 System Time: 2017-01-24T20:49:08.8436128Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by micdah** [comment]
I have tested the sample code on my Mac running the newest version of docker as well, and the issue doesn't seem to appear there. Although I am still experiencing intermitten connection failure across all threads at once - but here it recovers immediately and continues to work, no fuss.
Whereas on Windows, the port is dead and cannot be used again before Docker has been restarted.

**Comment by rneugeba** [comment]
@micdah Thanks for the very detailed repro (probably the most detailed I've seen sofar!).

Just for my record, I had to install the dotnet CLI tools from [here](https://www.microsoft.com/net/core#windowscmd) to compile and run the stress test.

The issue seems to be with a component called `VPNKit`. We run some stress tests against it in CI but your case seems to put an heavier load on it.

In the logs I see a lot of messages like this:
```
[21:42:21.456][VpnKit         ][Error  ] com.docker.slirp.exe: tcp:0.0.0.0:6379:tcp:172.19.0.2:6379 proxy failed with flow proxy a: write failed with Eof
```
and this:
```
[19:18:11.347][VpnKit         ][Error  ] com.docker.slirp.exe: Socket.Stream: caught A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.
```
and this:
```
[15:05:33.976][VpnKit         ][Error  ] com.docker.slirp.exe: Hvsock.read: An established connection was aborted by the software in your host machine.
```

With Beta 39, which has an updated version of VPNKit, I can also cause issue, albeit slightly different:
```
[13:49:44.377][VpnKit         ][Error  ] Process died
[13:49:49.381][VpnKit         ][Info   ] Starting C:\Program Files\Docker\Docker\Resources\com.docker.slirp.exe --ethernet hyperv-connect://acaef826-d7a2-41db-9685-1315c13f9a40 --port hyperv-connect://acaef826-d7a2-41db-9685-1315c13f9a40 --db \\.\pipe\dockerDataBase --debug --diagnostics \\.\pipe\dockerVpnKitDiagnostics
[13:49:49.383][VpnKit         ][Info   ] Started
[13:49:49.418][VpnKit         ][Info   ] com.docker.slirp.exe: Logging to stdout (stdout:true DEBUG:false)
[13:49:49.418][VpnKit         ][Info   ] com.docker.slirp.exe: Setting handler to ignore all SIGPIPE signals
[13:49:49.418][VpnKit         ][Info   ] com.docker.slirp.exe: vpnkit version %VERSION% with hostnet version  %HOSTNET_PINNED% uwt version 0.0.3 hvsock version 0.13.0 %HVSOCK_PINNED%
[13:49:49.418][VpnKit         ][Info   ] com.docker.slirp.exe: starting port forwarding server on port_control_url:hyperv-connect://acaef826-d7a2-41db-9685-1315c13f9a40 vsock_path:
[13:49:49.418][VpnKit         ][Info   ] com.docker.slirp.exe: connecting to acaef826-d7a2-41db-9685-1315c13f9a40:0B95756A-9985-48AD-9470-78E060895BE7
[13:49:49.421][VpnKit         ][Info   ] com.docker.slirp.exe: connecting to acaef826-d7a2-41db-9685-1315c13f9a40:30D48B34-7D27-4B0B-AAAF-BBBED334DD59
[13:49:49.421][VpnKit         ][Error  ] com.docker.slirp.exe: While watching /etc/resolv.conf: ENOENT
[13:49:49.421][VpnKit         ][Info   ] com.docker.slirp.exe: hosts file has bindings for 
[13:49:49.422][VpnKit         ][Info   ] com.docker.slirp.exe: hvsock connected successfully
[13:49:49.422][VpnKit         ][Info   ] com.docker.slirp.exe: hvsock connected successfully
[13:49:49.422][VpnKit         ][Info   ] com.docker.slirp.exe: attempting to reconnect to database
[13:49:49.422][DataKit        ][Info   ] com.docker.db.exe: accepted a new connection on \\.\pipe\dockerDataBase
[13:49:49.422][DataKit        ][Info   ] com.docker.db.exe: Using protocol TwoThousandU msize 8215
[13:49:49.426][VpnKit         ][Info   ] com.docker.slirp.exe: reconnected transport layer
[13:49:49.426][VpnKit         ][Info   ] com.docker.slirp.exe: remove connection limit
[13:49:49.432][VpnKit         ][Info   ] com.docker.slirp.exe: allowing binds to any IP addresses
[13:49:49.433][VpnKit         ][Info   ] com.docker.slirp.exe: updating resolvers to nameserver 8.8.8.8#53
[13:49:49.433][VpnKit         ][Info   ] order 0
[13:49:49.433][VpnKit         ][Info   ] nameserver 8.8.4.4#53
[13:49:49.433][VpnKit         ][Info   ] order 0
[13:49:49.433][VpnKit         ][Info   ] com.docker.slirp.exe: Add(3): DNS configuration changed to: nameserver 8.8.8.8#53
[13:49:49.433][VpnKit         ][Info   ] order 0
[13:49:49.433][VpnKit         ][Info   ] nameserver 8.8.4.4#53
[13:49:49.433][VpnKit         ][Info   ] order 0
[13:49:49.435][VpnKit         ][Info   ] com.docker.slirp.exe: Creating slirp server pcap_settings:disabled peer_ip:192.168.65.2 local_ip:192.168.65.1 domain_search: mtu:8000
[13:49:49.435][VpnKit         ][Info   ] com.docker.slirp.exe: PPP.negotiate: received ((magic VMN3T)(version 13)(commit"\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000"))
[13:49:49.436][VpnKit         ][Info   ] com.docker.slirp.exe: TCP/IP ready
[13:49:49.436][VpnKit         ][Info   ] com.docker.slirp.exe: stack connected
[13:49:49.436][VpnKit         ][Info   ] com.docker.slirp.exe: no introspection server requested. See the --introspection argument
[13:49:49.436][VpnKit         ][Info   ] com.docker.slirp.exe: starting diagnostics server on: \\.\pipe\dockerVpnKitDiagnostics
```
So, this actually crashes the process and the restart does not seem to work.

On a subsequent run (after restart) I got:
```
[13:58:45.123][VpnKit         ][Error  ] com.docker.slirp.exe: Hvsock.read: Lwt_stream.Closed
[13:58:48.312][VpnKit         ][Error  ] com.docker.slirp.exe: Hvsock.read: Lwt_stream.Closed
[13:58:48.312][VpnKit         ][Error  ] com.docker.slirp.exe: Socket.Stream: caught Lwt_stream.Closed
[13:59:30.685][VpnKit         ][Info   ] Thread 2950 killed on uncaught exception Assert_failure("src/core/lwt.ml", 497, 9)
[13:59:30.685][VpnKit         ][Info   ] Raised at file "uwt_preemptive.ml", line 384, characters 23-26
[13:59:30.685][VpnKit         ][Info   ] Called from file "lwt_unix/lwt_hvsock_main_thread.ml", line 49, characters 6-206
[13:59:30.685][VpnKit         ][Info   ] Called from file "thread.ml", line 39, characters 8-14
```

With beta38, I got a different crash:
```
Version: 1.13.0-beta38 (9805)
Channel: Beta
Sha1: 9c31a154d11ccf6a29f009610453eab4921bc6e8
[...]
[13:23:50.020][VpnKit         ][Error  ] com.docker.slirp.exe: Lwt.async failure "Assert_failure src/core/lwt.ml:497:9": Raised at file "src/core/lwt.ml", line 497, characters 9-21
[13:23:50.020][VpnKit         ][Info   ] Called from file "src/core/lwt.ml", line 201, characters 8-15
[13:23:50.020][VpnKit         ][Info   ] Called from file "src/core/lwt.ml", line 201, characters 8-15
[13:23:50.020][VpnKit         ][Info   ] Called from file "src/core/lwt.ml", line 201, characters 8-15
[13:23:50.020][VpnKit         ][Info   ] Called from file "src/core/lwt.ml", line 201, characters 8-15
[13:23:50.020][VpnKit         ][Info   ] Called from file "src/core/lwt.ml", line 305, characters 2-34
[13:23:50.020][VpnKit         ][Info   ] Called from file "uwt.ml", line 1669, characters 23-44
[13:23:50.020][VpnKit         ][Info   ] 
```

**Comment by micdah** [comment]
@rneugeba Ah yes, I forgot to mention that you needed .net core tools, that is my bad. :-) 
As this is such an intermittent issue, I figured a consistent way of reproducing it would be useful in diagnosing and identifying the issue.
Also I was curious whether the issue was specific to one of the services we are developing (normally these don't run under Docker, just for development at the moment), or if I could reproduce it with the most basic of application code.

Very interesting findings you have made regarding the probably component - curious that the logs show different parts failing in different versions.

**Comment by rneugeba** [comment]
The VPNKit code received significant updates since 1.13.0 stable both in the DNS handling code as well as a on the main data path (including some performance improvements) so I'm not surprised that it fails in different ways.

Your test code seems to put a considerable stress on it...If I understand the code, it tries to continuously open up to 500 connections (depending on the number entered) connections to container running redis and then continuously issues requests to the container.

As mentioned, we have some similar style tests in our regression suite, but they are slightly different

**Comment by micdah** [comment]
Yes that is correct, my first instinct what that it was related to either (or both) the number of concurrent connections and their activity/throughput.

So my first attempt to reproduce the issue, was having a configurable number of continuously open connections with high activity. 

In my example code, I make no attempt to re-connect if a connection fails - so the connections seem to be stable, until all connectivity fails at once for all open connections.

My findings are that the number of connections doesn't seem to affect if the issue can occur, but does seem to prolong the wait before the issue occurs. So my expectation would be, that a very specific timing between two or more concurrent connections seem to trigger the issue, thus the more concurrent connections and the more concurrent activity there are, the more likely the issue is to occur.

But to me, of course, this is just one big black box - so I might be totally off track. 😄

**Comment by atmorell** [comment]
I am seeing exactly the same behavior. VPNkit crashes under heavy load, and all containers stops responding. Restarting Docker makes everything great again ;) 

`
dmesg on the docker VM in Hyper-V shows the following error:
docker run --rm --privileged debian:jessie dmesg
[ 90.463682] device veth719b2e2 entered promiscuous mode
[ 90.463788] IPv6: ADDRCONF(NETDEV_UP): veth719b2e2: link is not ready
[ 90.463788] docker0: port 3(veth719b2e2) entered blocking state
[ 90.463788] docker0: port 3(veth719b2e2) entered forwarding state
[ 90.464119] docker0: port 3(veth719b2e2) entered disabled state
[ 90.539736] IPVS: Creating netns size=2104 id=12
[ 90.539739] IPVS: ftp: loaded support on port[0] = 21
[ 90.630341] eth0: renamed from vethc8fadb9
[ 90.680467] IPv6: ADDRCONF(NETDEV_CHANGE): veth719b2e2: link becomes ready
[ 90.680498] docker0: port 3(veth719b2e2) entered blocking state
[ 90.680498] docker0: port 3(veth719b2e2) entered forwarding state
[ 167.381158] docker0: port 4(vethb7059d0) entered blocking state
[ 167.381160] docker0: port 4(vethb7059d0) entered disabled state
`

Docker log:
https://paste.ee/p/FC1Sl

**Comment by dgageot** [comment]
@djs55 Could you please take a look at the diagnostics?

**Comment by fc0712** [comment]
I'm experiencing the same 

Will include log when I'm back at my Pc

**Comment by fc0712** [comment]
Any status on this issue?

**Comment by rneugeba** [comment]
we are still debugging the issue

**Comment by atmorell** [comment]
Is there any workaround for this crash?

**Comment by micdah** [comment]
Not to be pushy, but any updates on this issue?

If there is anything I can do to help (_such as testing beta/development versions_), I'll be happy to oblige.

**Comment by atmorell** [comment]
Almost thought it was solved by the latest update, but my containers became unresponsible after 1-2 hours of high load.

**Comment by rneugeba** [comment]
Unfortunately, the main developer for the component in question, VPNKit, was busy with some other work but will take a look this week. @micdah your reproduction is very useful for the debug. Again thanks for the effort in putting this together.

**Comment by djs55** [comment]
Sorry for the delay -- I'm planning to investigate this issue this week.

I notice in @rneugeba 's logs quoted above an assertion failed in `lwt`/`uwt` (the libraries we use on top of `libuv` for asychronous I/O on Windows and Mac) There is a newer version of `uwt` containing various bug fixes -- I'll see if that makes a difference first since I was planning an upgrade anyway.

The Mac and Windows versions of `com.docker.slirp.exe` process differ slightly in how they talk to the VM -- on the Mac we use a Unix domain socket while on Windows we use Hyper-V sockets. The Hyper-V socket code happens to stress the `Uwt_preemptive` module quite heavily which I notice is mentioned in @rneugeba 's quoted traceback. It may be possible to redesign that code to avoid that module if it turns out to be buggy.

I'll keep you informed of progress!

**Comment by djs55** [comment]
I've been running the repro case on beta build number `10183` which was briefly released this week and then recalled (due to discovering a serious bug elsewhere). The test has been running for several hours with no problems. I'll leave it over the weekend to see what happens. The main significant change in this build is the `libuv`/`uwt` upgrade.

**Comment by micdah** [comment]
How did it go with stress testing beta build 10183? Very excited to hear news on this issue. 😄

**Comment by micdah** [comment]
I have just updated my docker installation and tried to run my repro again, sadly it still seems to trigger.

This time though, at first I was able to restart the docker container without error (_previously I got an exception when it tried to bind to the same port when starting the container again_), but restarting the container doesn't fix the connectivity - it remains unreachable after the crash.

Output from `docker -v`:
```
Docker version 17.03.0-ce, build 60ccb22
```

After having restarted the container and tried running my program again to check if the container became reachable (_and verifying that it remained unreachable_), I tried restarting the container once more (_using `docker-compose restart` each time_), I received a new error I haven't seen before:
```
Error response from daemon: An operation on a socket could not be performed because the system lacked sufficient buffer space or because a queue was full.
```

It seems like the upgrade of VPNKit have had an effect on the issue, but doesn't seem to have fixed it.

**Comment by micdah** [comment]
Just tried to reproduce it once more, just to verify, this time it happened again (_both times, it seemed to trigger more quickly than it did previously, but that might just be imagination or random luck_).

This time, after re-running the program to verify the container is unreachable, and then trying to restart the container (`docker-compose restart`), I received the same error message as originally:
```
$ docker-compose restart
Restarting dockerbomb_redis_1 ... error

ERROR: for dockerbomb_redis_1  Cannot restart container 96579f13ce0f4322817342186728a646e3a33af3e7b6c91ead73a6becafb0742: driver failed programming external connectivity on endpoint dockerbomb_redis_1 (64d3c0dada115d7f483974802d85f3a8ff5851e4469b9c61db63c4cd0c4ae646): Error starting userland proxy: mkdir /port/tcp:0.0.0.0:6379:tcp:172.19.0.2:6379: input/output error
```

Odd that it let me restart the container, first time around, but not this time. Might just be due to the randomness of the undefined state it might be in, when the bug triggers.

**Comment by rneugeba** [comment]
@micdah thanks for testing. is there anything in the logs? Look for the ones prefixed with `vpnkit`.

**Comment by micdah** [comment]
@rneugeba Just re-ran the test again, so I could pinpoint the exact part of the log relevant to the point when the error occurs.

This is what was output, just when the error triggered:
```
[19:36:07.862][VpnKit         ][Error  ] Process died
[19:36:12.862][VpnKit         ][Info   ] Starting C:\Program Files\Docker\Docker\Resources\com.docker.slirp.exe --ethernet hyperv-connect://f08ee3f6-1270-41b8-8bd4-1874485c066a --port hyperv-connect://f08ee3f6-1270-41b8-8bd4-1874485c066a --db \\.\pipe\dockerDataBase --debug --diagnostics \\.\pipe\dockerVpnKitDiagnostics
[19:36:12.863][VpnKit         ][Info   ] Started
[19:36:12.883][VpnKit         ][Info   ] com.docker.slirp.exe: Logging to stdout (stdout:true DEBUG:false)
[19:36:12.883][VpnKit         ][Info   ] com.docker.slirp.exe: Setting handler to ignore all SIGPIPE signals
[19:36:12.883][VpnKit         ][Info   ] com.docker.slirp.exe: vpnkit version c41f7c8589352c95b14de636c895e8fbd72222e5 with hostnet version   uwt version 0.0.3 hvsock version 0.13.0 
[19:36:12.883][VpnKit         ][Info   ] com.docker.slirp.exe: starting port forwarding server on port_control_url:hyperv-connect://f08ee3f6-1270-41b8-8bd4-1874485c066a vsock_path:
[19:36:12.883][VpnKit         ][Info   ] com.docker.slirp.exe: connecting to f08ee3f6-1270-41b8-8bd4-1874485c066a:0B95756A-9985-48AD-9470-78E060895BE7
[19:36:12.885][VpnKit         ][Info   ] com.docker.slirp.exe: connecting to f08ee3f6-1270-41b8-8bd4-1874485c066a:30D48B34-7D27-4B0B-AAAF-BBBED334DD59
[19:36:12.885][VpnKit         ][Error  ] com.docker.slirp.exe: While watching /etc/resolv.conf: ENOENT
[19:36:12.885][VpnKit         ][Info   ] com.docker.slirp.exe: hosts file has bindings for client.openvpn.net openvpn-client.vpn.firstagenda.com
[19:36:12.886][VpnKit         ][Info   ] com.docker.slirp.exe: hvsock connected successfully
[19:36:12.886][VpnKit         ][Info   ] com.docker.slirp.exe: hvsock connected successfully
[19:36:12.886][VpnKit         ][Info   ] com.docker.slirp.exe: attempting to reconnect to database
[19:36:12.886][DataKit        ][Info   ] com.docker.db.exe: accepted a new connection on \\.\pipe\dockerDataBase
[19:36:12.886][DataKit        ][Info   ] com.docker.db.exe: Using protocol TwoThousandU msize 8215
[19:36:12.888][VpnKit         ][Info   ] com.docker.slirp.exe: reconnected transport layer
[19:36:12.888][VpnKit         ][Info   ] com.docker.slirp.exe: remove connection limit
[19:36:12.898][VpnKit         ][Info   ] com.docker.slirp.exe: allowing binds to any IP addresses
[19:36:12.900][VpnKit         ][Info   ] com.docker.slirp.exe: updating resolvers to nameserver 192.168.1.1#53
[19:36:12.900][VpnKit         ][Info   ] order 0
[19:36:12.900][VpnKit         ][Info   ] com.docker.slirp.exe: Add(3): DNS configuration changed to: nameserver 192.168.1.1#53
[19:36:12.900][VpnKit         ][Info   ] order 0
[19:36:12.902][VpnKit         ][Info   ] com.docker.slirp.exe: Creating slirp server peer_ip:192.168.65.2 local_ip:192.168.65.1 domain_search: mtu:8000
[19:36:12.902][VpnKit         ][Info   ] com.docker.slirp.exe: PPP.negotiate: received ((magic VMN3T)(version 13)(commit"\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000\000"))
[19:36:12.902][VpnKit         ][Info   ] com.docker.slirp.exe: TCP/IP ready
[19:36:12.902][VpnKit         ][Info   ] com.docker.slirp.exe: stack connected
[19:36:12.902][VpnKit         ][Info   ] com.docker.slirp.exe: no introspection server requested. See the --introspection argument
[19:36:12.902][VpnKit         ][Info   ] com.docker.slirp.exe: starting diagnostics server on: \\.\pipe\dockerVpnKitDiagnostics
```

**Comment by atmorell** [comment]
Crashing after a few minutes with the new build. Trying to start anything gives the following error:

Error response from daemon: driver failed programming external connectivity on endpoint sonarr (06921e1a5b924edb2347fde1b1e0ed18707e2508cb0052e090dbcaae930fa05b): Error starting userland proxy: mkdir /port/tcp:0.0.0.0:8989:tcp:172.17.0.4:8989: input/output error
Error: failed to start containers: sonarr

**Comment by micdah** [comment]
Yeah, today while trying to demonstrate a development environment I have been trying out for the past few months on my machine, to a few colleagues, Docker simply wouldn't start - it sat for a few minutes before showing an error message and the icon became red.

Sadly I was in a hurry, so I didn't have time to try and find something in the logs - and just tried starting docker again and now it works (_the machine have been restarted_).

Of course, these things will happen, but it is the first time I have ever experienced that failure.

**Comment by bluevulpine** [comment]
I've got a 'me, too!' on this.

I'm running Minecraft in a docker container so there's a lot of network activity out to the four people connected and this is happening 2-3 times a night when we're all connected. When this happens everyone gets booted, so it's easy to know when to go check the logs. The docker logs look the same as micdah - VpnKit posts a process died message, then proceeds through its restart. I can usually restore things to working condition by triggering a complete docker restart.

This is occurring with Docker Community Edition, Version 17.03.0-ce-win1 (10296), Channel: stable, e5a07a1, running on Windows 10 version 1607 build 14393.693.

**Comment by atmorell** [comment]
Any progress @djs55?

**Comment by micdah** [comment]
Hey, it's been a couple of weeks now - don't want to pressure anybody, but an update on this issue would be greatly appreciated. :-)

**Comment by friism** [comment]
@micdah as a workaround until this is fixed, can you run your integration tests on a stand-alone linux-machine or on a vm created with `docker-machine`. These won't exhibit the problem.

    docker-machine create -d hyperv --hyperv-virtual-switch <some-virtual-switch-you-created>

**Comment by SC7639** [comment]
I am having the same issue with my nginx container not accepting connections (seems to have frozen) and when I try to restart the container either using ```docker start vin_web_1``` or ```docker-compose up``` I get the following error:
```
Error response from daemon: driver failed programming external connectivity on endpoint vin_web_1 (013d29d8164ed5655c4b193c0198a8e9c4cafb163912326f2c19887f1d51b79e): Error starting userland proxy: mkdir /port/tcp:0.0.0.0:80:tcp:172.18.0.3:80: input/output error
Error: failed to start containers: vin_web_1
```

docker version info:
```
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:40:59 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   3a232c8
 Built:        Tue Feb 28 07:52:04 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

I've also uploaded a diagnostic with id: 7CE00045-6608-4DA4-9E8C-4479902B87FA/2017-03-21_11-50-37

If any other information is required to help with this issue let me know and I will provide it.

Restarting docker for windows does solve the issue but it would be nice to not have to wait for docker to for windows to reboot while trying to work.

**Comment by Naragato** [comment]
Myself and several colleagues are experiencing the same issue.  Restarting docker then the containers temporarily resolves the issue.

`Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:40:59 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   3a232c8
 Built:        Tue Feb 28 07:52:04 2017
 OS/Arch:      linux/amd64
 Experimental: true`

A diagnostic was uploaded with id: D5E01A54-0F2C-479B-8E3B-F5CD022D8D92/2017-03-22_14-47-30

**Comment by AngryBeaver** [comment]
same here

**Comment by mittork** [comment]
Any progress @djs55 ?

**Comment by tparikka** [comment]
I have also been encountering this problem. Restarting Docker and the containers has worked for me as well. I am using Docker for end-to-end web app UI tests with a single container running selenium/hub:3.2.0 and ten containers running selenium/node-chrome:3.2.0. spun up with docker-compose. The hub container seems to be the one that bails out, and it tends to do so between 5 and 15 minutes from the time I start running the test fixtures.

> Client:
>  Version:      17.03.1-ce
>  API version:  1.27
>  Go version:   go1.7.5
>  Git commit:   c6d412e
>  Built:        Tue Mar 28 00:40:02 2017
>  OS/Arch:      windows/amd64
> 
> Server:
>  Version:      17.03.1-ce
>  API version:  1.27 (minimum version 1.12)
>  Go version:   go1.7.5
>  Git commit:   c6d412e
>  Built:        Fri Mar 24 00:00:50 2017
>  OS/Arch:      linux/amd64
>  Experimental: true

I've uploaded diagnostic 7EBC5B9F-3979-420B-B185-E677DE90285E/2017-03-31_17-22-24.

**Comment by Naragato** [comment]
Still no-one assigned to the bug, it's pretty bad.  The issue has existed for quite a while now....

**Comment by bvitale** [comment]
We were unable to adopt Docker for Windows due to this issue.  Had to revert to a VM based approach.

**Comment by markopadjen** [comment]
Fix this pls.

**Comment by tparikka** [comment]
@rneugeba @djs55 Is it possible to provide an update on progress, even if it's just that it is stalled waiting on resources to become available to investigate? The last real update I see was a suggested workaround by @friism 27 days ago, but there hasn't been any response on the submitted diagnostics since March 2.

**Comment by jeanlaurent** [comment]
Hi everyone,

Sorry for the lack of response, we don't have a clear answer yet, @djs55 tried the reproduce the example which ran for several days without seeing the problem.

We're going to re-run that on a spare system, setup for that issue specifically for a longer time again and see what comes out of it.

We will publish the result in here.

Thanks for your patience.

**Comment by micdah** [comment]
@jeanlaurent From your last comment, it sounded like the reproduction didn't work anymore (being able to run for days without encountering the issue).

So to check whether the reproduction code was still relevant, I tried re-running it with my current version of Docker for windows (latest). Tried a total of three times, and the issue occurred each time.

I noticed that it seemed more "random" when it would hit the issue, first run ran for about a minute before dying. Second time took well over 10 minutes, and third time it failed even before all threads were up and running.

Each time I ran with 500 threads.

**Comment by narkedboy** [comment]
Same issue occurs for me, running a single cloudera quickstart docker container, under medium network load, all network ports die within half hour to a few hours. The error messages in the log files are the same as those reported.

**Comment by djs55** [comment]
@micdah sorry for the delay getting back to you. I've been running the example program with 500 threads for a little over 24 hours now with no sign of a problem. I'm using the latest edge version:
```
Version 17.05.0-ce-rc1-win8 (11189)
Channel: edge
73d01bb
```
on Windows version 1607 (OS Build 14393.953).

Could you upgrade to the latest edge version, reproduce the problem and upload a fresh diagnostic?

**Comment by SC7639** [comment]
I've found that it only happens when my cpu is at 100% for more than a couple of hours. I found what was causing the massive cpu usage and I've not had the issue since

**Comment by tparikka** [comment]
@djs I installed the edge release and ran through my full suite of Selenium tests and did not encounter failures, so it seems good _so far_ on my end.

**Comment by micdah** [comment]
@djs55 I've just updated my docker to edge
```
# docker -v
Docker version 17.05.0-ce-rc1, build 2878a85
```

Running my example code, after approx 10 minutes, all the connections died. Here is a diagnostic id
```
30667474-C49F-4185-B957-3A7AE1F38393/2017-04-21_07-52-01
```

Restarting the container, doesn't do anything either. A restart of docker is still needed.

**Comment by djs55** [comment]
I still see an error like this in the logs:
```
[07:49:21.959][VpnKit         ][Error  ] vpnkit.exe: Lwt.async failure "Assert_failure src/core/lwt.ml:500:9": Raised at file "src/core/lwt.ml", line 500, characters 9-21
[07:49:21.960][VpnKit         ][Info   ] Called from file "src/core/lwt.ml", line 202, characters 8-15
[07:49:21.960][VpnKit         ][Info   ] Called from file "src/core/lwt.ml", line 202, characters 8-15
[07:49:21.960][VpnKit         ][Info   ] Called from file "src/core/lwt.ml", line 202, characters 8-15
[07:49:21.960][VpnKit         ][Info   ] Called from file "src/core/lwt.ml", line 202, characters 8-15
[07:49:21.960][VpnKit         ][Info   ] Called from file "src/core/lwt.ml", line 207, characters 8-15
[07:49:21.960][VpnKit         ][Info   ] Called from file "src/core/lwt.ml", line 290, characters 6-42
[07:49:21.960][VpnKit         ][Info   ] Called from file "uwt.ml", line 1798, characters 23-44
```
Do you know if all network traffic stops, or is it only ports? I notice these errors too:
```
[07:51:35.026][VpnKit         ][Error  ] vpnkit.exe: Upstream DNS server 8.8.4.4:53 has dropped 6 packets in a row: assuming it's offline
[07:51:35.026][VpnKit         ][Error  ] vpnkit.exe: Upstream DNS server 8.8.8.8:53 has dropped 6 packets in a row: assuming it's offline
[07:51:35.026][VpnKit         ][Error  ] vpnkit.exe: Upstream DNS server 192.168.0.2:53 has dropped 6 packets in a row: assuming it's offline
```
-- this could be real or it could be a cascade failure.

**Comment by micdah** [comment]
It was only the ports that stopped, I was working on the machine while running the test and I didn't notice any network outage.

I did however have an active VPN connection to our Amazon environment, but only traffic to/from our servers are routed through that, the rest goes straight out the house.

**Comment by Bazmcl** [comment]
seem to be suffering from the same issue - not opening a lot of ports but a lot of traffic

Uploaded diagnostic EB0E3EF6-6538-4B43-ADB9-8CFE688B9BBF/2017-05-04_09-26-08

**Comment by tparikka** [comment]
@djs55 Has there been any new information unearthed since @micdah confirmed it was only the ports that stopped in his environment?

**Comment by bvitale** [comment]
A diagnostic was uploaded with id: 0BFBE273-80BD-45FB-AD1A-36352590E93D/2017-05-16_13-44-49

```
[13:41:50.640][VpnKit         ][Warning] vpnkit.exe: ARP table has no entry for 172.18.0.2
[13:41:50.640][VpnKit         ][Error  ] vpnkit.exe: PPP.listen callback caught Ipv4.Make(Ethif)(Arpv4).Routing.No_route_to_destination_address(_)
[13:42:02.762][VpnKit         ][Error  ] vpnkit.exe: Hvsock.shutdown_write: got Eof
[13:42:02.762][VpnKit         ][Error  ] vpnkit.exe: tcp:0.0.0.0:5000:tcp:172.18.0.15:3000 proxy failed with flow proxy b: write failed with Eof
[13:42:02.970][VpnKit         ][Error  ] vpnkit.exe: Socket.Stream: caught An existing connection was forcibly closed by the remote host.
```

Windows 10 14393.1066

My scenario is sending a large number of requests to the same port.

**Comment by tparikka** [comment]
@djs55 or @jeanlaurent Has there been any progress on this issue, and any more tests/diagnostics we can provide that would assist at this time? Thank you for your help!

**Comment by mattjanssen** [comment]
Same issue with Win 10 Pro and Docker Edge. No matter what random (unused) port combinations I used, I got the same error.

I fixed it by stopping Docker from the system tray and restarting it. After I fixed the problem I created diagnostic 950F6894-7F6D-4081-BDCE-7B35E19A391B/2017-05-30_16-55-11.

```
C:\Users\Matt>docker run -p "10392:13293" agaveapi/beanstalkd-console
docker: Error response from daemon: driver failed programming external connectivity on endpoint 
hopeful_mcnulty (fa135ff9192e4bd4f103e5f6128863d174b426483463d32558f332440d5865a4): 
Error starting userland proxy: 
mkdir /port/tcp:0.0.0.0:10392:tcp:172.17.0.2:13293: input/output error.
```

**Comment by alarys** [comment]
Hi,
I'm experiencing similar problems as those above.
I submitted a diagnostic
A diagnostic was uploaded with id: 7FF77AD4-6196-4C0C-BF18-962C00826605/2017-06-06_14-05-31

My Docker version 17.03.1-ce, build c6d412e

I tried to use the latest Edge release, but my docker configurations throw errors when I try to create containers. It seems to be complaining about local drive mappings. Not sure what is going on there.

My containers are not responding as above. My use is downloading content from Usenet. As soon as a download begins, and the data transfer ramps up, all my containers stop responding. Restarting docker gets things working again.

I have mitigated the problem somewhat by throttling bandwidth. But even throttling bandwidth to 500 kbps, the problem still resurfaces after a while. I can reliably reproduce by not throttling the bandwidth and kicking off a download.

I'm really quite disappointed with how docker on windows is handling large data throughput. And this seems like an issue that others have experienced too.

**Comment by tparikka** [comment]
@djs55 @jeanlaurent There hasn't been any input from the Docker team on this issue since April 21. I'm hopeful that if there hasn't been any progress that the community may be able to help by trying out builds or providing additional diagnostics. Thank you!

**Comment by mikesnare** [comment]
@djs55 @jeanlaurent It's now been close to 4 months with no developer response to what could be argued is a pretty serious -- crippling -- bug.  Any updates?  I'm using Docker for Windows to spin up a zookeeper and a couple kafka instances and it dies pretty consistently under load and then fails during restart with the same errors others are describing, forcing me to restart docker entirely.

**Comment by agentilela** [comment]
Same issue here, uploaded diagnostic: `0CC3ABDF-040B-4BF0-9D39-B24CAE24F6ED/2017-08-10_19-47-42`

Here's the interesting stuff

```
[19:37:29.451][VpnKit         ][Info   ] Tcp.PCB: ERROR: thread failure; terminating threads and closing connection

[19:37:29.452][VpnKit         ][Error  ] vpnkit.exe: Lwt.async failure (Invalid_argument Lwt.wakeup_result): Raised at file "format.ml", line 241, characters 41-52

[19:37:29.452][VpnKit         ][Info   ] Called from file "format.ml", line 482, characters 6-24

[19:37:29.452][VpnKit         ][Info   ] 

[19:40:29.855][VpnKit         ][Info   ] Tcp.Segment: TCP retransmission on timer seq = 531271309

[19:40:31.855][VpnKit         ][Info   ] Tcp.Segment: TCP retransmission on timer seq = 531271309

[19:40:35.855][VpnKit         ][Info   ] Tcp.Segment: TCP retransmission on timer seq = 531271309

[19:40:43.855][VpnKit         ][Info   ] Tcp.Segment: TCP retransmission on timer seq = 531271309

[19:40:59.857][VpnKit         ][Info   ] Tcp.Segment: TCP retransmission on timer seq = 531271309

[19:41:07.180][VpnKit         ][Error  ] Process died
```

**Comment by jinxuunity** [comment]
I have the same issue in docker _version 17.06.1-ce-win24 (13025)_ , as welll as last version. 

when I execute docker-compose in powershell console, I saw 
`WindowsError: [Error 2] The system cannot find the file specified: u'***********************'
Failed to execute script docker-compose`

Have seen in docker daemon log.  
`[08:50:59.912][VpnKit         ][Error  ] vpnkit.exe: Hvsock.read: An established connection was aborted by the software in your host machine.
`

**Comment by smellinet** [comment]
Hello,
I have the same issue with  latest version :
Version 17.06.2-ce-win27 (13194)
Channel: stable
428bd6c
After a heavy load , the network of container is broken.
 the stop/start of container doesn't solve the problem:
```
Error response from daemon: driver failed programming external connectivity on endpoint tapo (23bc1c5ec134f7b164eb6c35e810cd89e876d8c8da3b46db4d8685b642f8ac8d): Error starting userland proxy: mkdir /port/tcp:0.0.0
.0:5500:tcp:172.17.0.2:5500: input/output error
```
diagnostic id upload : 
Diagnostics successfully uploaded (C64A9176-3C73-4FBC-B4FA-D4B0017B689C/2017-09-07_10-18-23).

**Comment by Naragato** [comment]
I can't believe this still isn't a priority to fix. :(

**Comment by mittork** [comment]
@djs55 , can you provide any ETA for this to be fixed? So far Docker for Windows is not usable in a productive way for us and we have to think about workarounds (like using another standalone Host and configure docker to client connect to this).

But I ask: How can I trust a software for production, which is not able to handle a bit more load in development stage. I know it is related to the VPN-kit, but anyway....

**Comment by tparikka** [comment]
@djs55 @jeanlaurent Is there any more information available on this problem, an ETR, or even an updated priority?

**Comment by TheFamilyRoom** [comment]
We are experiencing the same behavior. heavy load on single port and the docker bridge falls over. the containers are still running but can't be accessed. it seems like this is not a priority for anyone to address but it is holding us up. proposed solurions:

run on mac/linux - we will try this next
run less load? - sorta defeats the point. 

anyone else have success getting this to work on Win 10?

**Comment by micdah** [comment]
Yeah I have more or less given up on running heavy loads on Docker for Windows, interestingly I don't seem to have the same issues after we are moving our services over onto Kubernetes running via minikube on windows.

Naturally this environment is just an extra stack on top of Docker, but it seems like Minikube at least, runs "better" on Windows (_using Hyper-V, but it is also possible to use VirutalBox_).

**Comment by SC7639** [comment]
I'm still experiencing this issue now and again. It happened today and I had to restart docker for windows for a container to use the port again.

**Comment by tparikka** [comment]
@djs55, @jeanlaurent, can you comment on whether or not this issue been officially abandoned?

**Comment by djs55** [comment]
@tparikka We've not abandoned the issue, but unfortunately other issues have been higher priority recently -- I apologise for the delay.

We're hoping to update the version of the Linux kernel we use to 4.14, which has a newer implementation of Hyper-V sockets which we use for exposing ports. We should be able to drop some of the workarounds for bugs in the previous version and hopefully this will make the whole system more reliable. As part of this update we'll do some general stress testing and attempt to reproduce this issue.

Thanks again for your patience.

**Comment by SC7639** [comment]
Thanks for the update

**Comment by Michal-Svoboda** [comment]
We suffer the same issue in our project as well.
@djs55 - I would like to ask you, if there is any schedule when there will be the new version of Docker available using the newer implementation of Hyper-V sockets?

And what is the current status of this issue?

**Comment by vohtaski** [comment]
Same problem here. Running MariaDb in a docker container on Windows.
After several thousand requests, it dies with "dial tcp 127.0.0.1:33061: getsockopt: connection refused"
Would be amazing to have a fix or a workaround

**Comment by tparikka** [comment]
@djs55 @jeanlaurent I wanted to check in on this since it's been about 4 months. Is there any update on this issue, and perhaps is there a separate Git issue that's been logged for the Linux kernel version update that you hope will improve stability under load so we can follow it?

**Comment by sw-carlin** [comment]
This problem seems to have improved in the stable channel, as I'm on Docker version 18.03.1-ce and am able to still run docker commands when the exposed ports of my containers aren't responsive; In the previous version that was not possible.

I am also able to recover from the situation by stopping some of the containers which I guess is freeing up frozen sockets? I'm running 20 containers that compose a microservice ecosystem with lots of traffic moving between them and can trigger the situation by running any of my system integration tests. I will try running the tests from inside the container composition to see if that is a good workaround.

**Comment by tparikka** [comment]
@djs55, @jeanlaurent it has been over 7 months since the last update. Is there any further information on this issue?

**Comment by djs55** [comment]
@tparikka sorry for the delay. There has been some progress: we've started updating the Hyper-V socket implementation used in several of the components to remove a complex (possibly buggy) workaround for bugs in old Windows builds (< 14393). Once this is done we'll update the Hyper-V socket GUIDs that we use and then we can bump the kernel version. These changes will be merged into the development branch gradually -- I'll let you know when there are interesting development builds you can test.

**Comment by tparikka** [comment]
@djs55 Has there been a release that we could play with to test for improvements?

**Comment by djs55** [comment]
There are some changes to port forwarding and Hyper-V sockets in today's stable release candidate build: https://download.docker.com/win/stable/29211/Docker%20for%20Windows%20Installer.exe -- this build is probably worth testing. Let me know if you get a chance to try it!

**Comment by tparikka** [comment]
I ran my Selenium tests successfully against the release candidate build a few times and it didn't blow up. Since I first posted to this issue though I've migrated my test assemblies to .NET Core 2.1 and the underlying framework to .NET Standard 2.0, so my test environment isn't quite the same as when I started looking at it. I'd be interested to hear if others also see the issue resolved - @micdah, @TheFamilyRoom, @smellinet, others who have reported the issue any chance you could also try the new build and let us know what results you see?

**Comment by JPMoresmau** [comment]
I see to get this issue only since recently (like update of Docker for Windows), never happened before.
Diagnostics FBC58536-F77C-4909-9BBE-918AA324B487/20181213174347

Symptoms are the same: after a few minutes of activity containers are not reachable from localhost.

**Comment by llenrup** [comment]
I only recently started getting related issue see #3108 
was working before last update

**Comment by tparikka** [comment]
It took longer after the prerelease provided by @djs55, but I have run into the same issue again:

ERROR: for selenium-hub  Cannot start service selenium-hub: driver failed programming external connectivity on endpoint selenium-hub (7f8d436acd812ee3d7ed9e96f1591f5b2fcb0882adb095feeb02bd4e861342ad): Error starting userland proxy: mkdir /port/tcp:0.0.0.0:4444:tcp:172.19.0.2:4444: input/output error
ERROR: Encountered errors while bringing up the project.

Docker Engine 18.09.0
Compose: 1.23.2
Docker Desktop 2.0.0.0-win81
Windows 10 Build 1809 x64

**Comment by djs55** [comment]
I've fixed a number of bugs in the port-forwarding code which should make it more stable after load. If you'd like to try an early version of them I have put links to development builds here: https://github.com/docker/for-win/issues/3257#issuecomment-461563065

Let me know if this makes things any better. Thanks for your patience with this issue!

**Comment by tparikka** [comment]
@djs55 I can't run a deployment build but I have switched to the edge channel on my local instance to look for updates. Which edge release do you anticipate will get the update, or can you post here when it goes live?

**Comment by tparikka** [comment]
I ran into the issue on this 2.0.2.1-Edge release:

Creating selenium-hub ... error

ERROR: for selenium-hub  Cannot start service selenium-hub: driver failed programming external connectivity on endpoint selenium-hub (c90bbfd6b8f3f449da754229ffd9082e1d28112dc28c0c835ec70f591ae36ef3): Error starting userland proxy: mkdir /port/tcp:0.0.0.0:4444:tcp:172.18.0.2:4444: input/output error

ERROR: for selenium-hub  Cannot start service selenium-hub: driver failed programming external connectivity on endpoint selenium-hub (c90bbfd6b8f3f449da754229ffd9082e1d28112dc28c0c835ec70f591ae36ef3): Error starting userland proxy: mkdir /port/tcp:0.0.0.0:4444:tcp:172.18.0.2:4444: input/output error
ERROR: Encountered errors while bringing up the project.

Engine 18.09.2
Compose 1.24.0-rc1
Version 2.0.2.1 31274 edge

**Comment by bennettellis** [comment]
Running a commercial piece of software in a container, exercising the REST API repeatedly. Really frustrating bug. I hit the API maybe 50 times and down goes the port. Recreating the docker container doesn't fix, so it's not the http server on the container causing it (e.g. rate limiting). Disabling firewall, anti-virus etc doesn't fix, so they don't seem to be an issue. Only fix is to restart docker for windows. Oddly enough, the host still reports through PortQryV2 that the localhost port is still being listened to, but any attempt to go the next network layer to communicate with the http endpoint in the container just hangs. Extremely consistent behavior. Very frustrating. I may go ahead and try #3257 fix to see if it helps with my issue and help get this tested and released. Thanks!

**Comment by bennettellis** [comment]
Just moved to edge v 2.0.3.0 (31778) and voila no more port hanging for me at least. Docker --version reports Docker version 18.09.3, build 774a1f4 for the record. Seems fixed. Thanks!!

**Comment by michaeladada** [comment]
Since the issue seems to be with the vpnkit, you can bypass it and connect directly to the MobyLinuxVM.
Instead of "localhost" use "10.0.75.2".
This is the default IP assigned to the MobyLinuxVM.
You can see what the IP is by running the following commands:
```
docker run -it --privileged --pid=host justincormack/nsenter1 /bin/sh
ifconfig
```
The relevant IP is the one assigned to **hvint0**

**Comment by Koricz** [comment]
Hi, I have similar problem.

Windows 10 local installation: Apache, PHP, MySQL database, Elasticsearch, Rabbit MQ
Linux docker containers: Elasticsearch, Rabbit MQ

Results when running RabbitMQ consumer written in PHP which should index cca 50 000 objects / rows from MySQL to Elasticsearch database:

1) Local installation only (without Docker) - All 50 000 objects are processed
2) With actual release Docker version - cca 400 request processed then Docker must be restarted to accept any new incoming connections.
3) With Docker 2.0.4.1 (34207) edge - cca 400 request processed, then connection is reset / closed => PHP script is terminated, but containers still accepts any new incoming connection and no Docker restart is needed. 

In log - v2.0.4.1 (34207) there are messages like: 

```
[00:12:50.383][ApiProxy       ][Info   ] time="2019-05-17T00:12:50+02:00" msg="proxy << GET /v1.25/containers/808a870d6de247ccce06f46e797e49e90bf979cd7e4aff411571636c5e3ca6a2/json (2.0005ms)\n"
[00:12:50.384][ApiProxy       ][Info   ] time="2019-05-17T00:12:50+02:00" msg="proxy << GET /v1.25/containers/8dd4675e193b48d41c8062da9cc493a87ec5c70d77654949960b4727420770eb/json (2.0034ms)\n"
[00:35:21.999][VpnKit         ][Error  ] vpnkit.exe: tcp:0.0.0.0:9200:tcp:172.19.0.3:9200 proxy failed with flow proxy a: attempted to write to a closed flow
[00:36:46.562][VpnKit         ][Error  ] vpnkit.exe: tcp:0.0.0.0:5672:tcp:172.19.0.4:5672 proxy failed with flow proxy a: attempted to write to a closed flow
[00:47:20.992][VpnKit         ][Error  ] vpnkit.exe: Socket.tcp:127.0.0.1:56170.write TCPv4: caught An established connection was aborted by the software in your host machine.
[00:47:20.992][VpnKit         ][Info   ]  returning Eof
```

**Comment by tparikka** [comment]
I encountered this defect again today:
ERROR: for selenium-hub  Cannot start service selenium-hub: driver failed programming external connectivity on endpoint selenium-hub (e1d864b70783f0b77693bd56cb43ca9176531b72979993550d303f73c143571b): Error starting userland proxy:
ERROR: Encountered errors while bringing up the project.

@djs55 Can you speak to the ongoing issues folks are having?

**Comment by tparikka** [comment]
/remove-lifecycle stale

@djs55, anything here?

**Comment by tparikka** [comment]
Bumping this, @djs55.

/remove-lifecycle stale

**Comment by djs55** [comment]
Does anyone have a current repro case that works with a recent stable version of Docker Desktop that they could share with me?

Thanks in advance!

**Comment by tparikka** [comment]
I had run into it a while ago with Selenium but it's a super unreliable reproduction vector. I tried taking the DockerBomb project and updating it to .NET Core 3.1 but I'm not super versed in Redis so I'm not sure I'm using it right to try and reproduce the issue. @micdah I don't suppose you'd be able to take a look? I did push my revision to the project up to github.com/tparikka/DockerBomb.

**Comment by micdah** [comment]
@tparikka Just merged your fork into my repo and verified that the code compiles and runs. Alas I am not anymore working from a Windows machine so I can’t say whether the program will still show the bug or not - but for anyone running on windows, they could try it out with a few thousand “bombs” and see if it still suddenly dies as it did over three years ago.

To run the code, in short, do:
```
docker-compose up -d
dotnet run —project DockerBomb
```

**Comment by tparikka** [comment]
@djs55 I just ran the updated DockerBomb app on my machine (i5-7600K OC@3.8GHz, 16 GB RAM) and despite maxing out my CPU and hitting 3000 threads connecting to a Redis container in Docker I wasn't able to reproduce the issue. It seems to be working on my end. I'm inclined to suggest we leave this issue open long enough for other more recent participants (such as @Koricz) to respond if they have a reproducible scenario for this issue, and let the stalebot close it out if no one responds.

**Comment by jhnns** [comment]
@tparikka What Docker engine version did you use?

**Comment by tparikka** [comment]
@jhnns I'm on Docker Desktop 2.2.0.4 Stable Engine 19.03.8 on Windows 10 Pro Version 1909.

**Comment by jk2K** [comment]
I encountered the same problem, I solved it, you can try adjusting `open files` and `max user processes`
```bash
# The maximum number of open file descriptors
ulimit -Sn 65535
# The maximum number of processes available to a single user
ulimit -Su 100000
```

**Comment by azumukupoe** [comment]
I think WSL 2 kind of solved this
